Adding  ZDC Unpacker to L1RawtoDigi

matt2275 opened the pull request:
#### PR description:

Adding a ZDC unpacker for the L1 Emulation

This is a work in progress Andrew David Loeliger and Dinyar Rabady  are more expect in this area and will help solve the issues with the PR

All file changes are in EventFilter/L1TRawToDigi/plugins/implementations_stage2/
2 files were created 
ZDCUnpacker.cc
ZDCUnpacker.h

3 Files Changed:
GTSetup.cc
GTCollections.cc
GTCollections.h

Currently there is a complier error I'm not sure how to resolve:
  /afs/cern.ch/user/m/mnickel/private/ZDC_Trig/New_WF/CMSSW_13_2_0_pre3/src/EventFilter/L1TRawToDigi/plugins/implementations_stage2/GTCollections.h:40:33: error: 'l1t::EtSumBxCollection* l1t::stage2::GTCollections::getZDCSums(unsigned int)' marked 'override', but does not override
     40 |       inline EtSumBxCollection* getZDCSums(const unsigned int copy) override { return zdcsums_[copy].get(); };



#### PR validation:
No validation has been done yet.

matt2275 committed:
Adding ZDCUnpacker files

matt2275 committed:
Update GTSetup.cc for ZDCUnpacker

matt2275 committed:
Update ZDCUnpacker.cc to  fix double variable declarations

matt2275 committed:
Update ZDCUnpacker.cc to clarify ZDC data format

matt2275 committed:
Update GTSetup.cc

Added Et sum for ZDC unpacker in registerProducts

matt2275 committed:
Update GTCollections.h

added zdc to GTCollections

matt2275 committed:
Update GTCollections.cc

added ZDCSum to GTCollections

matt2275 committed:
Update ZDCUnpacker.cc

Updating names due to GTCollection changes

matt2275 committed:
Update ZDCUnpacker.h

Updated names due to GTCollection changes

matt2275 committed:
Update GTSetup.cc

matt2275 committed:
Update GTSetup.cc

matt2275 committed:
Merge pull request #1 from matt2275/master

Merging CMS-SW master to ZDCUnpacker

bundocka commented on EventFilter/L1TRawToDigi/plugins/implementations_stage2/ZDCUnpacker.cc:
```diff
@@ -0,0 +1,71 @@
+    ZDCUnpacker::ZDCUnpacker() : ZDCSumCopy_(0) {}
+
+    bool ZDCUnpacker::unpack(const Block& block, UnpackerCollections* coll) {
+      using namespace l1t::stage2::layer2;
```
this should just be `l1t::stage2`, since the ZDC is not part of Calo Layer 2

matt2275 replied:
I addressed this in my commit, Fixing Issues for PR 42733 (First Attempt). I had to change demux::nOutputFramePerBX to layer2::demux::nOutputFramePerBX to solve compiler errors. The ZDC data has 6 frames for BX which is the value of nOutputFramePerBX but this may be coincident.

dinyar replied:
Yeah, I was unsure whether to flag the use of `demux::nOutputFramePerBX` in my first pass. I'd say this could technically change and become inconsistent with the number of input frames to the uGT, but in practice I have a hard time imagining how that would do. (So the bottom line from my side was that it's probably reasonable. If I remember correctly we just put the literal 6 in the muon unpacker which is also not nice.. ) @bundocka may have better thought through input than me, though.

bundocka replied:
demux::nOutputFramePerBX == zdc::nOutputFramePerBX is a coincidence, I would not recommend using the demux constant, but instead create zdc::nOutputFramePerBX, for clarity. as Dinyar said, this should be fine as it is, and it shouldn't change, but the two numbers are technically unrelated so they shouldn't use the same constant.

bundocka replied:
I tell a lie, it is not a coincidence at all :)  6bx is the maximum length allowed for the GT input packet. the input links are clocked at 240 MHz. 240 / 6 = 40 MHz.

matt2275 replied:
To add the zdc::nOutputFramePerBX variable. should I add it to the L1TStage2Layer2Constants.cc file or create a new constants file only for zdc.  I could also add it to the ZDCUnpacker.h but I don't like that since I believe the variable will also be used by the zdc packer in the future. 

 I would have the variable be this in all cases

 const unsigned int l1t::stage2::zdc::nOutputFramePerBX = 6;

bundocka replied:
I would be tempted to put it somewhere like  GTSetup.h, since it is a general parameter for the input to the GT. (Then technically this parameter should be used throughout the muon and layer 2 unpackers)
https://github.com/cms-sw/cmssw/blob/master/EventFilter/L1TRawToDigi/plugins/implementations_stage2/GTSetup.h

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42733/36824

ERROR: Build errors found during clang-tidy run.
```
EventFilter/L1TRawToDigi/plugins/implementations_stage2/ZDCUnpacker.cc:33:61: error: no member named 'getZDCSums' in 'l1t::stage2::L1TObjectCollections'; did you mean 'getEtSums'? [clang-diagnostic-error]
      auto res_ = static_cast<L1TObjectCollections*>(coll)->getZDCSums(ZDCSumCopy_);
                                                            ^~~~~~~~~~
                                                            getEtSums
--
EventFilter/L1TRawToDigi/plugins/implementations_stage2/GTCollections.h:40:69: error: only virtual member functions can be marked 'override' [clang-diagnostic-error]
      inline EtSumBxCollection* getZDCSums(const unsigned int copy) override { return zdcsums_[copy].get(); };
                                                                    ^~~~~~~~~
Suppressed 871 warnings (871 in non-user code).
--
EventFilter/L1TRawToDigi/plugins/implementations_stage2/GTCollections.h:40:69: error: only virtual member functions can be marked 'override' [clang-diagnostic-error]
      inline EtSumBxCollection* getZDCSums(const unsigned int copy) override { return zdcsums_[copy].get(); };
                                                                    ^~~~~~~~~
Suppressed 977 warnings (977 in non-user code).
--
gmake: *** [config/SCRAM/GMake/Makefile.coderules:129: code-checks] Error 2
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
```

missirol commented on EventFilter/L1TRawToDigi/plugins/implementations_stage2/GTCollections.cc:
```diff
@@ -9,13 +9,15 @@ namespace l1t {
       event_.put(std::move(muonShowers_[0]), "MuonShower");
       event_.put(std::move(egammas_[0]), "EGamma");
       event_.put(std::move(etsums_[0]), "EtSum");
+      event_.put(std::move(zdcsums_[0]), "ZDCSum");
```
Please change this string to `EtSumZDC`, here and elsewhere in this PR (see https://github.com/cms-sw/cmssw/pull/42707#discussion_r1314495801).

matt2275 replied:
This was addressed in Fixing Issues for PR 42733 (First Attempt). I did not update the zdcsums since I thought changing to etsumzdcs would be confusing but I can change that as well

missirol commented on EventFilter/L1TRawToDigi/plugins/implementations_stage2/GTSetup.cc:
```diff
@@ -55,6 +56,7 @@ namespace l1t {
       prod.produces<MuonShowerBxCollection>("MuonShower");
       prod.produces<EGammaBxCollection>("EGamma");
       prod.produces<EtSumBxCollection>("EtSum");
+      prod.produces<EtSumBxCollection>("ZDCSum"); // added addition EtSum collection for ZDC  unpacker 
```
> added addition EtSum collection for ZDC  unpacker

Please improve the comment (or remove it).

missirol commented on EventFilter/L1TRawToDigi/plugins/implementations_stage2/GTSetup.cc:
```diff
@@ -97,6 +102,7 @@ namespace l1t {
       muon_unp->setMuonCopy(amc - 1);
       egamma_unp->setEGammaCopy(amc - 1);
       etsum_unp->setEtSumCopy(amc - 1);
+      zdc_unp->setZDCSumCopy(amc - 1);
```
```suggestion
      zdc_unp->setEtSumZDCCopy(amc - 1);
```
I would also change the name of this method.

aloeliger commented:
@fwyzard & @Martin-Grunewald if you haven't seen this, I wanted to let you know this exists.

I see @missirol has already found it.

missirol commented:
Notifying also @mmusich as @cms-sw/hlt-l2.

---

> Currently there is a complier error I'm not sure how to resolve:
> /afs/cern.ch/user/m/mnickel/private/ZDC_Trig/New_WF/CMSSW_13_2_0_pre3/src/EventFilter/L1TRawToDigi/plugins/implementations_stage2/GTCollections.h:40:33: error: 'l1t::EtSumBxCollection* l1t::stage2::GTCollections::getZDCSums(unsigned int)' marked 'override', but does not override
> 40 | inline EtSumBxCollection* getZDCSums(const unsigned int copy) override { return zdcsums_[copy].get(); };


I would guess one needs to add a method here:
https://github.com/cms-sw/cmssw/blob/630332245c94bf42249109d922061741ed59d2f6/EventFilter/L1TRawToDigi/plugins/implementations_stage2/L1TObjectCollections.h#L28

aloeliger commented:
@dinyar In case you were unaware that this has been opened, I also wanted you to know that this is available on github for review.

dinyar commented on EventFilter/L1TRawToDigi/plugins/implementations_stage2/ZDCUnpacker.cc:
```diff
@@ -0,0 +1,71 @@
+        
+        l1t::EtSum zdcm{l1t::EtSum::kZDCM};
+        zdcm.setHwPt(raw_data & 0xFFFF);
+        zdcm.setP4(l1t::CaloTools::p4Demux(&zdcm));
```
Just to check here: my understanding is that this function computes the four vector with the implicit understanding that the LSB is 0.5 GeV. Is that the case for the ZDC sums too? (Again this is based on me not finding documentation, so may be a moot point.)

icali replied:
it is correct your interpretation but in ZDC we should use the numerical expression and not the conversion in energy as it is going to be wrong, in my understanding. @cfmcginn can comment.

dinyar replied:
Right, I think the uGT emulator always uses the hw values anyway and gets the scales from the menu XML. The physical pT should, however be correct because that's what people usually look at in e.g. the ntuples.

fwyzard replied:
what about the ZDC L1 candidate used in the object map ?

fwyzard replied:
what about the ZDC L1 candidate used in the object map ?

dinyar replied:
Yes, absolutely. If someone accesses the physical pT value of that object they would get a value of `hw_code*0.5` GeV. I'm not entirely sure from Ivan's answer whether this is correct.

dinyar commented on EventFilter/L1TRawToDigi/plugins/implementations_stage2/GTSetup.cc:
```diff
@@ -116,6 +122,7 @@ namespace l1t {
         res[16] = tau_unp;
         res[18] = tau_unp;
         res[20] = etsum_unp;
+        res[22] = zdc_unp;
```
I suppose it's confirmed that ZDC inputs come on link 11 of the uGT?

icali replied:
I understood it was link 71 (0x8E)

matt2275 replied:
I was not sure of the link. If the link is 71, then it should be res[142]?

bundocka replied:
If the logical link ID is 71, then yes I think res[142] should be correct. 
Cross referencing the GT input data format [1] with the unpacker [2] I think supports this. 

[1] https://twiki.cern.ch/twiki/bin/viewauth/CMS/GlobalTriggerInputDataFormat
[2] https://github.com/cms-sw/cmssw/blob/440318ca7e69a9a79fcd262a9deafb54bb933916/EventFilter/L1TRawToDigi/plugins/implementations_stage2/GTSetup.cc#L114

bundocka replied:
(I think even links is Rx and odd links is Tx)

dinyar replied:
Yeah, my understanding is also that input ports get even numbers, so it would be 142 if it's link 71.

icali commented on EventFilter/L1TRawToDigi/plugins/implementations_stage2/ZDCUnpacker.cc:
```diff
@@ -0,0 +1,71 @@
+        uint32_t raw_data = block.payload().at(iFrame +1); // ZDC - info is found on frame 1 of each bx
+        
+        l1t::EtSum zdcm{l1t::EtSum::kZDCM};
+        zdcm.setHwPt(raw_data & 0xFFFF);
```
The sum should be 10 bits if I'm not wrong but @jmmans can comment as for what Herbert says it is an exact copy of what the uHTR sends. As far as I know frame 1 is ZDC- and frame 2 is ZDC+

dinyar replied:
Ok, in that case:
```suggestion
        zdcm.setHwPt(raw_data & 0x3FF);

```

(but as mentioned below I would put that into a named constant)

matt2275 replied:
To close this comment, I will update the bit mask and put it into a variable. Are we sure bit shifting is not required to extract the zdc info (i.e. rawdata >> 8) ?

aloeliger replied:
@matt2275 My understanding of Herbert on the email was the uGT stores these in the same word format that ZDC hands them off to uGT. Does ZDC store bits unrelated to the overall HW Pt that will need to be masked or shifted away?

bundocka replied:
I don't think there is any data other than the energy count (even so, it should always be masked to the know width). 
ps. I found some doc, a search for ZDC gives quite a lot of useful info for this PR:
https://github.com/cms-l1-globaltrigger/mp7_ugt_legacy/blob/master/doc/mp7_ugt_firmware_specification/pdf/gt-mp7-firmware-specification.pdf

matt2275 committed:
Fixing Issues for PR 42733 (First Attempt)

Addressing some of the issues raised by cms-sw

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42733/36825

- This PR adds an extra 32KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42733/36825/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42733/36825/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

bundocka commented on EventFilter/L1TRawToDigi/plugins/implementations_stage2/ZDCUnpacker.cc:
```diff
@@ -0,0 +1,71 @@
+        LogDebug("L1T") << "ZDC -: pT " << zdcm.hwPt() << " bx " << bx;
+
+        res_->push_back(bx, zdcm);
+        
```
I think the EtSums are selected using the sumType Enum (rather than index in the EtSum vector):
https://github.com/cms-sw/cmssw/blob/master/L1Trigger/L1TGlobal/src/EnergySumZdcCondition.cc#L139
https://github.com/cms-sw/cmssw/blob/master/L1Trigger/L1TGlobal/src/EnergySumCondition.cc#L125

matt2275 committed:
Fixing Issues for PR 42733 (Second Attempt)

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42733/36827

- This PR adds an extra 20KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42733/36827/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42733/36827/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

aloeliger commented:
@matt2275 could you apply the auto-formatting to this and once you've handled all the comments move it off of draft and to an active PR? @elfontan Do we have any data with ZDC in the uGT input record? I think we're still in need of a proper test before this becomes official or HLT has to test/rely on it.

bundocka commented:
@icali should confirm but I think run 373061 should have ZDC data in the RAW if all went to plan. 
looking in OMS the ZDC FEDs were included: 1136 (ZDC), 11136 (ZDC TCDS)

aloeliger commented:
Okay. If that exists then there is a path to testing this. @matt2275, can you:

1. Run a quick technical check to make sure this unpacker can unpack those runs and produces an output collection
2. Run a quick sanity check on the output sums to make sure they aren't zero, or some maxed out integer Pt-wise
3. Run a technical check to make sure that the gt emulator can be run off of this unpacked collection

?

matt2275 commented:
> Okay. If that exists then there is a path to testing this. @matt2275, can you:
> 
> 1. Run a quick technical check to make sure this unpacker can unpack those runs and produces an output collection
> 2. Run a quick sanity check on the output sums to make sure they aren't zero, or some maxed out integer Pt-wise
> 3. Run a technical check to make sure that the gt emulator can be run off of this unpacked collection
> 
> ?

To Clarify should I do the testing first or make it an active PR first?

aloeliger commented:
Neither precludes the other.

matt2275 committed:
Fixing Issues for PR 42733 ( Third Attempt)

Applied cmsbuild formatting and declared zdc::nOutputFramesperBX in GTSetup.cc instead of GTSetup.h

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42733/36831

- This PR adds an extra 20KB to repository

cmsbuild commented:
A new Pull Request was created by  @matt2275  for master.

It involves the following packages:

- EventFilter/L1TRawToDigi (**l1**)


@epalencia, @cmsbuild, @aloeliger can you please review it and eventually sign? Thanks.
@dinyar, @missirol, @Martin-Grunewald, @thomreis, @eyigitba this is something you requested to watch as well.
@rappoccio, @antoniovilela, @sextonkennedy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

aloeliger commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-49d39c/34667/summary.html
**COMMIT**: 15fc44bd9cbe42a9189f908b62d653d029a6f7f2
**CMSSW**: CMSSW_13_3_X_2023-09-08-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/42733/34667/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially removed 5 lines from the logs
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 337 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3153414
 - DQMHistoTests: Total failures: 146
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3153246
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 207 log files, 159 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_3_X_2023-09-08-1100+49d39c/58787/triggerResults): no differences found

mandrenguyen commented:
Perhaps someone can summarize the status here?

fwyzard commented on EventFilter/L1TRawToDigi/plugins/implementations_stage2/ZDCUnpacker.cc:
```diff
@@ -0,0 +1,71 @@
+      LogDebug("L1T") << "Block ID  = " << block.header().getID() << " size = " << block.header().getSize();
+
+      int nBX =
+          int(ceil(block.header().getSize() / zdc::nOutputFramePerBX));  // Since there 6 frames per demux output event
```
This looks like a division between two integers, `lock.header().getSize()` and `zdc::nOutputFramePerBX`.
If this is indeed the case, the result is automatically truncated towards zero, and the use of the `ceil()` function will not do anything.
If the intent is to round the result of the division _up_, it should be written as
```suggestion
          int(std::ceil(static_cast<float>(block.header().getSize()) / zdc::nOutputFramePerBX));
          // Since there 6 frames per demux output event
```
or
```suggestion
          (block.header().getSize() + zdc::nOutputFramePerBX - 1) / zdc::nOutputFramePerBX;
          // Since there 6 frames per demux output event

```

aloeliger replied:
Not that this excuses it, but I believe this is how it is done in the current EtSum unpacker: https://github.com/cms-sw/cmssw/blob/99f2750c476c243522d4599697496b1af4c29ce7/EventFilter/L1TRawToDigi/plugins/implementations_stage2/EtSumUnpacker.cc#L20-L21

fwyzard commented on EventFilter/L1TRawToDigi/plugins/implementations_stage2/ZDCUnpacker.cc:
```diff
@@ -0,0 +1,71 @@
+      int firstBX = -(ceil((double)nBX / 2.) - 1);
+      int lastBX;
+      if (nBX % 2 == 0) {
+        lastBX = ceil((double)nBX / 2.);
```
It was just checked that `nBX` is even, so there is no need to cast to `double` and call `ceil`.
Just
```suggestion
        lastBX = nBX / 2;
```

fwyzard commented on EventFilter/L1TRawToDigi/plugins/implementations_stage2/ZDCUnpacker.cc:
```diff
@@ -0,0 +1,71 @@
+      if (nBX % 2 == 0) {
+        lastBX = ceil((double)nBX / 2.);
+      } else {
+        lastBX = ceil((double)nBX / 2.) - 1;
```
here `nBX` is odd, so this is equivalent to
```suggestion
        lastBX = nBX / 2;

```

fwyzard commented on EventFilter/L1TRawToDigi/plugins/implementations_stage2/ZDCUnpacker.cc:
```diff
@@ -0,0 +1,71 @@
+
+      // Find the central, first and last BXs
+      int firstBX = -(ceil((double)nBX / 2.) - 1);
+      int lastBX;
+      if (nBX % 2 == 0) {
+        lastBX = ceil((double)nBX / 2.);
+      } else {
+        lastBX = ceil((double)nBX / 2.) - 1;
+      }
```
In fact, this whole block can be simplified to
```suggestion
      int lastBX = nBX / 2;
```

matt2275 replied:
I can apply these changes but these lines were copied from the etsumunpacker (https://github.com/matt2275/cmssw/blob/ZDCUnpacker/EventFilter/L1TRawToDigi/plugins/implementations_stage2/EtSumUnpacker.cc#L24).

fwyzard replied:
Please fix the new code in this PR.
A separate PR can be made to fix the other instances.

fwyzard commented on EventFilter/L1TRawToDigi/plugins/implementations_stage2/ZDCUnpacker.cc:
```diff
@@ -0,0 +1,71 @@
+      // expect the first four frames to be the first 4 EtSum objects reported per event (see CMS IN-2013/005)
+
+      // Find the central, first and last BXs
+      int firstBX = -(ceil((double)nBX / 2.) - 1);
```
```suggestion
      int firstBX = (nBX / 2) - nBX + 1;
```

fwyzard commented on EventFilter/L1TRawToDigi/plugins/implementations_stage2/ZDCUnpacker.h:
```diff
@@ -0,0 +1,23 @@
+    class ZDCUnpacker : public Unpacker {
+    public:
+      ZDCUnpacker();
+      ~ZDCUnpacker() override{};
```
```suggestion
      ~ZDCUnpacker() override = default;

```

fwyzard commented on EventFilter/L1TRawToDigi/plugins/implementations_stage2/ZDCUnpacker.h:
```diff
@@ -0,0 +1,23 @@
+
+      bool unpack(const Block& block, UnpackerCollections* coll) override;
+
+      virtual inline void setEtSumZDCCopy(const unsigned int copy) { EtSumZDCCopy_ = copy; };
```
why `virtual` ?

matt2275 replied:
It's virtual since the similar function in etsumunpacker was virtual.

fwyzard replied:
But `setEtSumZDCCopy()` is implemented only in a single class - there is no inheritance involved.
Please remove the `virtual` function specifier here.
A separate PR can be made to fix the existing code.

matt2275 committed:
Fixing Issues for PR 42733 ( 4th Attempt)

Simplified first and last BX calculation and removed unnecessary virtual on setEtSumZDCCopy

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42733/36864

- This PR adds an extra 20KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42733/36864/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42733/36864/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

matt2275 committed:
Fixing Issues for PR 42733 ( 5th Attempt)

Fixing formatting for cmsbuild

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42733/36866

- This PR adds an extra 40KB to repository

cmsbuild commented:
Pull request #42733 was updated. @epalencia, @cmsbuild, @aloeliger can you please check and sign again.

fwyzard commented on EventFilter/L1TRawToDigi/plugins/implementations_stage2/ZDCUnpacker.cc:
```diff
@@ -0,0 +1,65 @@
+      uint32_t zdc_mask = 0x3FF;
+      LogDebug("L1T") << "Block ID  = " << block.header().getID() << " size = " << block.header().getSize();
+
+      int nBX = int(ceil(block.header().getSize() / zdc::nOutputFramePerBX));
```
Please, remove `int(ceil(`, since it doesn't do anything on the result of an integer division.

If you need to round the result up, please change it to
```suggestion
      int nBX = (block.header().getSize() + zdc::nOutputFramePerBX - 1) / zdc::nOutputFramePerBX;
```

fwyzard commented:
Question: do we have RAW data collected with an L1 menu that includes ZDC, that can be used to test this unpacker ?

fwyzard commented:
> Question: do we have RAW data collected with an L1 menu that includes ZDC, that can be used to test this unpacker ?

@missirol wrote me that we do:
### run 373272
```
/store/data/Run2023F/PPRefZeroBias0/RAW/v1/000/373/272/00000/459549ce-9e92-459e-8aff-a59ba58c5b0a.root
```
### run 373273
```
/store/data/Run2023F/PPRefZeroBias0/RAW/v1/000/373/273/00000/07524cfb-c470-4e7f-a207-0a0ebd54a0a1.root
/store/data/Run2023F/PPRefZeroBias0/RAW/v1/000/373/273/00000/0b67bed8-4282-4e64-89c5-da0aabad5408.root
/store/data/Run2023F/PPRefZeroBias0/RAW/v1/000/373/273/00000/31244cc8-2458-40df-8674-cda977bb6760.root
/store/data/Run2023F/PPRefZeroBias0/RAW/v1/000/373/273/00000/62b6625f-4e95-4cd7-b569-6a66118770eb.root
/store/data/Run2023F/PPRefZeroBias0/RAW/v1/000/373/273/00000/aa69e356-12e7-47d2-87f9-075bff7dbb78.root
/store/data/Run2023F/PPRefZeroBias0/RAW/v1/000/373/273/00000/d87dc8a1-5eb9-4705-9853-524353cbef46.root
/store/data/Run2023F/PPRefZeroBias0/RAW/v1/000/373/273/00000/df26043a-5581-423f-ad31-1a8771d5d4f6.root
```

At CERN they can be found under `/eos/cms/tier0/store/...` for the time being.

Can anybody check that the ZDC unpacker + uGT emulator reproduces the online L1T decision ?

matt2275 commented:
> Okay. If that exists then there is a path to testing this. @matt2275, can you:
> 
> 1. Run a quick technical check to make sure this unpacker can unpack those runs and produces an output collection
> 2. Run a quick sanity check on the output sums to make sure they aren't zero, or some maxed out integer Pt-wise
> 3. Run a technical check to make sure that the gt emulator can be run off of this unpacked collection
> 
> ?

[1] seems to have been completed. I was able to produce an edm file with the zdc unpacker information.  The edm file is located here /afs/cern.ch/user/m/mnickel/public/ZDC_Trig/l1tCalo_2016_EDM_Normal.root.

 I've begun working on [2] but as it stands I think I think I'll need to create a new analyzer similar to (https://github.com/cms-sw/cmssw/blob/master/L1Trigger/L1TCalorimeter/python/l1tStage2CaloAnalyzer_cfi.py) for gtstage2Digis. If someone knows an easier way to fill a root file or histogram with unpacker information, that would be extremely helpful.

For [3], I'm not sure what the technical check entails so I will need help and guidance to complete that task.

matt2275 commented:
> Okay. If that exists then there is a path to testing this. @matt2275, can you:
> 
> 1. Run a quick technical check to make sure this unpacker can unpack those runs and produces an output collection
> 2. Run a quick sanity check on the output sums to make sure they aren't zero, or some maxed out integer Pt-wise
> 3. Run a technical check to make sure that the gt emulator can be run off of this unpacked collection
> 
> ?

[1] seems to have been completed. I was able to produce an edm file with the zdc unpacker information.  The edm file is located here /afs/cern.ch/user/m/mnickel/public/ZDC_Trig/l1tCalo_2016_EDM_Normal.root.

 I've begun working on [2] but as it stands I think I think I'll need to create a new analyzer similar to (https://github.com/cms-sw/cmssw/blob/master/L1Trigger/L1TCalorimeter/python/l1tStage2CaloAnalyzer_cfi.py)(https://github.com/cms-sw/cmssw/blob/master/L1Trigger/L1TCalorimeter/python/l1tStage2CaloAnalyzer_cfi.py) for gtstage2Digis. If someone knows an easier way to fill a root file or histogram with unpacker information, that would be extremely helpful.

For [3], I'm not sure what the technical check entails so I will need help and guidance to complete that task.

aloeliger commented:
> > Question: do we have RAW data collected with an L1 menu that includes ZDC, that can be used to test this unpacker ?

@fwyzard This is what I have been asking the ZDC team to reproduce with this:

> Okay. If that exists then there is a path to testing this. @matt2275, can you:
> 
> 1. Run a quick technical check to make sure this unpacker can unpack those runs and produces an output collection
> 2. Run a quick sanity check on the output sums to make sure they aren't zero, or some maxed out integer Pt-wise
> 3. Run a technical check to make sure that the gt emulator can be run off of this unpacked collection
> 
> ?

So far the ZDC team has demonstrated that the unpacker produces a collection properly. They are now looking at making sure the sums are made properly.

aloeliger commented:
My apologies for the slow responses to things here. I have been in Athens for the L1T workshop.

matt2275 committed:
Fixing ZDC res_ map

unpacking map for ZDC was 71 instead of 142

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42733/36893

- This PR adds an extra 44KB to repository

cmsbuild commented:
Pull request #42733 was updated. @epalencia, @cmsbuild, @aloeliger can you please check and sign again.

matt2275 commented:
Running with this file:  /store/data/Run2023F/PPRefZeroBias0/RAW/v1/000/373/273/00000/07524cfb-c470-4e7f-a207-0a0ebd54a0a1.root Most the digis are empty but the etsum  and etsumZDC digis are filled.  An edm file with 2k processed events is located here /afs/cern.ch/user/m/mnickel/public/ZDC_Trig/l1tCalo_2016_EDM_2k.root. The histogram shows the histogram etsumzdc.  I also attached a block with zdc info and the unpacker output. Here the words and the unpacker output match.

With this info, it seems the unpacker is working as expected and I'm moving on to the third check and final which is confirming the zdc unpacker works with the gt emulator. 

![image](https://github.com/cms-sw/cmssw/assets/54369650/0e34acfa-d0de-48aa-a580-c73ef0dc55d1)
[ZDC_Block with unpacking.txt](https://github.com/cms-sw/cmssw/files/12603276/ZDC_Block.with.unpacking.txt)

aloeliger commented:
Okay, @cms-sw/hlt-l2 & @fwyzard

I have conferred with the ZDC team and asked them to run 3 tests to demonstrate the basic function of the unpacker:

1. Run a quick technical check to make sure this unpacker can unpack runs with ZDC at uGT and produce an output collection
2. Run a quick sanity check on the output sums to make sure they aren't zero, or some maxed out integer Pt-wise
3. Run a technical check to make sure that the gt emulator can be run off of this unpacked collection

And, given the short time frame, I think they have done so. For each point in order

> 1. Run a quick technical check to make sure this unpacker can unpack runs with ZDC at uGT and produce an output collection

@matt2275 took a test configuration, ran the GT unpacker, and took a look at the output sequences:

The ZDC unpacker will produce an output sequence with GT:
![image](https://github.com/cms-sw/cmssw/assets/28607665/574a91c5-7c2f-4444-b52b-9e319b63630a)

> 2. Run a quick sanity check on the output sums to make sure they aren't zero, or some maxed out integer Pt-wise

matt has demonstrated that if you examine the output of the unpacker the sums are not _obviously_ / _egregiously_ bugged, which I think is as much as can be asked here:

![image](https://github.com/cms-sw/cmssw/assets/28607665/96b93ff6-57a8-4857-ab51-db4a6daf7cb2)

> 3. Run a technical check to make sure that the gt emulator can be run off of this unpacked collection

Rerunning the L1 analysis chain through to ntuples, and running GT off the ZDC unpacker, matt was also able to get the chain to run without crashing and examine the output, which is in agreement with the unpacker output from before:

![image](https://github.com/cms-sw/cmssw/assets/28607665/cb6bf4df-fa71-41c3-b4c8-15bd218afccf)



Strictly, I think this satisfies me and L1 given the urgency on this. I plan to sign on this in the next few hours after the tests pass unless I hear from you guys with a pressing reason this shouldn't proceed. I want to put out a final call for checking for any problems.

aloeliger commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-49d39c/34776/summary.html
**COMMIT**: 7422859b2180eadd8ec21cacb47a332260af0699
**CMSSW**: CMSSW_13_3_X_2023-09-14-2300/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/42733/34776/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially removed 6 lines from the logs
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 3348648
 - DQMHistoTests: Total failures: 3
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3348623
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 214 log files, 167 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_3_X_2023-09-14-2300+49d39c/58892/triggerResults): no differences found

fwyzard commented:
Sorry @aloeliger , I'm not able to follow part of your [last comment](https://github.com/cms-sw/cmssw/pull/42733#issuecomment-1720841194).

> 3. Run a technical check to make sure that the gt emulator can be run off of this unpacked collection

> Rerunning the L1 analysis chain through to ntuples, and running GT off the ZDC unpacker, matt was also able to get the chain to run without crashing and examine the output, which is in agreement with the unpacker output from before

does this last point mean that
  - the ZDC input to the uGT was unpacked from the uGT raw data using the module introduced in this PR (#42733)
  - the resulting collection was to the uGT emulator (which now includes #42634/#42706)
  - the results of the uGT emulator were compared with the uGT results from the data
?

fwyzard commented:
OK, so if I understand correctly you are able to run these two workflows

1. read the uGT results from data, and fill the ntuples with these original results

2. unpack all uGT inputs (including ZDC), run the uGT emulator on them, and fill the ntuples with the emulated results


Assuming that's correct, do you have a way to compare event by event the original and emulated results ?

If they are identical, everything is fine.

aloeliger commented:
+l1

- at this point I haven't heard any show stoppers

@cms-sw/orp-l2 when this is done, a new version of the software is likely needed urgently. I can discuss this at any point Monday if needed.

I think it worth noting that I am signing this commit from the Acropolis of Athens :)

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @rappoccio, @antoniovilela, @sextonkennedy (and backports should be raised in the release meeting by the corresponding L2)

mandrenguyen commented:
@matt2275 Can you please prepare the 13_2_X backport?

antoniovilela commented:
> OK, so if I understand correctly you are able to run these two workflows
> 
>     1. read the uGT results from data, and fill the ntuples with these original results
> 
>     2. unpack all uGT inputs (including ZDC), run the uGT emulator on them, and fill the ntuples with the emulated results
> 
> 
> Assuming that's correct, do you have a way to compare event by event the original and emulated results ?
> 
> If they are identical, everything is fine.

@matt2275 @aloeliger 
Given the urgency and validation performed so far, we will merge this PR. Please continue with the event-by-event validation proposed.

antoniovilela commented:
+1

matt2275 commented:
To do the backport should I do a similar procedure as this: https://github.com/cms-sw/cmssw/pull/42781 but add EventFilter/L1TRawtoDigi instead of L1Trigger/L1TGlobal

mandrenguyen commented:
I suppose, although I'm not sure sure what procedure was followed for this other PR.

Normally one should grab the latest CMSSW_13_2_X integration build, which at this moment you would do as follows: 
`cmsrel CMSSW_13_2_X_2023-09-15-1100`
`cd  CMSSW_13_2_X_2023-09-15-1100/src`
`cmsenv`

Ideally you would use the `git-cherry-pick` command with the hash of each commit, but you have many of them. 
In case you don't manage, I guess since the changes are modest perhaps in this case it would be enough to then do:
`git cms-addpkg EventFilter/L1TRawToDigi`
and then add your changes by hand **very carefully**

kpedro88 commented:
Repeated cherry-pick is not necessary. There are backporting instructions here: http://cms-sw.github.io/tutorial-resolve-conflicts.html#backporting-a-pr

To make a backport simpler, you can first make a copy of your 13_3_X branch and squash all of your commits down to one commit, and then just rebase that single commit.

fwyzard commented:
We should really move the default action in CMSSW to "Squash and merge" ...

mmusich commented:
> To make a backport simpler, you can first make a copy of your 13_3_X branch and squash all of your commits down to one commit, and then just rebase that single commit.

on a similar tone, I am not sure why this PR was accepted with 18 commits (most of them with uninformative messages). A minimum squash would have made life simpler.

mandrenguyen commented:
I just went ahead and made PR in 13_2_X after rebase/squash:
https://github.com/cms-sw/cmssw/pull/42800

@matt2275 (and others) please check.

antoniovilela commented:
> We should really move the default action in CMSSW to "Squash and merge" ...

Seems like something to consider.
@rappoccio
